### PR TITLE
Don't report numeric inference ambiguity when we have previous errors

### LINKED ISSUE #95648: Unrelated compiler errors affects inferring a type
When I run this code (via `cargo run`):

```rust
use rust_decimal::Decimal;

fn main() {
  let test = Decimal::from(42);
  println!("My number is {}", test);
}
```

I get the output I expected (i.e., a single line of output that says "My number is 42").

However, if I run this code (via the same `cargo run` with no other changes):

```rust
use rust_decimal::Decimal;

fn main() {
  foo(3);
  let test = Decimal::from(42);
  println!("My number is {}", test);
}
```

I get two compiler errors. The first error indicates that `foo` is not in scope - this makes sense, I have not defined `foo` anywhere. However, the second error indicates (provided below) does not make sense:

```
error[E0283]: type annotations needed
 --> src/main.rs:5:16
  |
5 |     let test = Decimal::from(42);
  |                ^^^^^^^^^^^^^ cannot infer type for type `{integer}`
  |
  = note: multiple `impl`s satisfying `Decimal: From<{integer}>` found in the `rust_decimal` crate:
          - impl From<i128> for Decimal;
          - impl From<i16> for Decimal;
          - impl From<i32> for Decimal;
          - impl From<i64> for Decimal;
          and 8 more
```

It appears that, when there are no other compiler issues, the compiler can infer a type for `42` just fine. However, as soon as any other compiler issue appears, it loses the ability to infer type. That doesn't make sense to me - the ability to infer type should be unrelated to any other compiler errors present.

### Meta

`rustc --version --verbose`:
```
rustc 1.58.1 (db9d1b20b 2022-01-20)
binary: rustc
commit-hash: db9d1b20bba1968c1ec1fc49616d4742c1725b4b
commit-date: 2022-01-20
host: x86_64-unknown-linux-gnu
release: 1.58.1
LLVM version: 13.0.0
```

### PR DESCRIPTION
**Pull request by compiler-errors**
Fixes #95648

### CODE REVIEW

File: src/test/ui/traits/test-2.stderr [hunk 1]
```diff
@@ -79,35 +79,7 @@ LL | trait bar { fn dup(&self) -> Self; fn blah<X>(&self); }
    = note: required because of the requirements on the impl of `CoerceUnsized<Box<dyn bar>>` for `Box<{integer}>`
    = note: required by cast to type `Box<dyn bar>`
 
-error[E0283]: type annotations needed
```
Comment (compiler-errors) [inline src/test/ui/traits/test-2.stderr:82]: hehe, wonderful to see this issue getting better too

**Comment by jackh726** [comment]
Not sure this is the best fix. It would be better, I think to fallback even under error.

**Comment by compiler-errors** [comment]
@jackh726, like in general? that would have a bunch of unintended side-effects i think -- the choice to not fallback when infcx is tainted by errors is intentional apparently: https://doc.rust-lang.org/stable/nightly-rustc/src/rustc_typeck/check/fallback.rs.html#92

**Comment by compiler-errors** [comment]
I'll investigate modifying the fallback logic to unconditionally apply float and integer fallback even when infcx is tainted, but i'm assuming that'll have additional diagnostic fallout that may need discussion on its merit, or extra patching-up to suppress, so i will keep that separate from this PR.

File: compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs [hunk 2]
```diff
@@ -2226,9 +2227,10 @@ impl<'a, 'tcx> InferCtxtPrivExt<'a, 'tcx> for InferCtxt<'a, 'tcx> {
         post.dedup();
 
         if self.is_tainted_by_errors()
-            && crate_names.len() == 1
-            && ["`core`", "`alloc`", "`std`"].contains(&crate_names[0].as_str())
-            && spans.len() == 0
+            && (crate_names.len() == 1
```
Comment (jackh726) [inline compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs:2230]: Could this check just be simplified to "if the self type is a numeric infer"?

Comment (compiler-errors) [reply compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs:2230]: So we need to check for int/float variables not just in self, but in the trait's substs as well, so we don't report the things that implement `ConcreteType: Trait<{integer}>`. 

Also removing the core/alloc/std crate name check leads to some regressions in errors reported with simd stuff and when we're unable to infer const variables due to other errors :/

**Pull request merged**